modules/plugins/languages/scala.nix

```diff
@@ -0,0 +1,138 @@
+              type = "scala",
+              request = "launch",
+              name = "RunOrTest",
+              metals = {
+                runType = "runOrTestFile",
+                --args = { "firstArg", "secondArg", "thirdArg" }, -- here just as an example
+              },
+            },
+            {
+              type = "scala",
+              request = "launch",
+              name = "Test Target",
+              metals = {
+                runType = "testTarget",
+              },
+            },
+          }
+        '';
+      };
+    };
+  };
+
+  config = mkIf cfg.enable (
+    mkMerge [
+      (mkIf cfg.treesitter.enable {
+        vim.treesitter.enable = true;
+        vim.treesitter.grammars = [cfg.treesitter.package];
+      })
+      (mkIf (cfg.lsp.enable || cfg.dap.enable) {
+        vim = {
+          startPlugins = ["nvim-metals"];
+          pluginRC.nvim-metals = entryAfter ["lsp-setup"] ''
+            local metals_caps = capabilities  -- from lsp-setup
+
+            local attach_metals_keymaps = function(client, bufnr)
+              attach_keymaps(client, bufnr)  -- from lsp-setup
+              vim.api.nvim_buf_set_keymap(bufnr, 'n', '${cfg.lsp.extraMappings.listCommands}', '<cmd>lua ${listCommandsAction}<CR>', {noremap=true, silent=true, desc='Show all Metals commands'})
+            end
+
+            metals_config = require('metals').bare_config()
+            ${optionalString usingLualine "metals_config.init_options.statusBarProvider = 'on'"}
+
+            metals_config.capabilities = metals_caps
+            metals_config.on_attach = function(client, bufnr)
+              ${optionalString usingDap "require('metals').setup_dap()"}
+              attach_metals_keymaps(client, bufnr)
+            end
+            metals_config.settings = {
+               metalsBinaryPath = "${cfg.lsp.package}/bin/metals",
+               showImplicitArguments = true,
+               showImplicitConversionsAndClasses = true,
+               showInferredType = true,
+               excludedPackages = {
+                 "akka.actor.typed.javadsl",
+                 "com.github.swagger.akka.javadsl"
+               },
+
+            }
+
+            metals_config.handlers["textDocument/publishDiagnostics"] = vim.lsp.with(
+              vim.lsp.diagnostic.on_publish_diagnostics, {
+                virtual_text = {
+                  prefix = '⚠',
+                }
+              }
+            )
+
+            -- without doing this, autocommands that deal with filetypes prohibit messages from being shown
+            vim.opt_global.shortmess:remove("F")
+
+            local lsp_group = vim.api.nvim_create_augroup('lsp', { clear = true })
+
+            vim.api.nvim_create_autocmd('FileType', {
+                group = lsp_group,
+                pattern = {'java', 'scala', 'sbt'},
+                callback = function()
+                    require('metals').initialize_or_attach(metals_config)
+                end,
```

Comment:
This entire section feels really off to me. I can't tell you right off the bat what would be better (if that's possible) but we are hardcoding *a lot* of values here with no good way of overriding.


To remedy this, we could:

a. try defining config options in lua tables with a distinctive prefix (e.g. `__nvf_metals_config`) and explicitly instruct users to modify said table to override in Lua configurations
*or*

b. Define most of those under a specific set under `vim.languages.scala`, say, `languages.scala.lsp.extraOpts` which gets passed to `toLuaObject from extended library.

Pick your poison.

Reply:
I went for an `extraSettings` option that is set as `metals_config.settings`. Not sure if that solves all your concerns, though. I saw that there's `mkPluginSetupOption` and I was inclined to use it, but decided against as I am not technically passing the table to `setup`.